src/server.rs

```diff
@@ -704,9 +704,16 @@ async fn handle_login<C: Config>(
             .chain(&server.0.public_key_der)
             .finalize();
 
-        let hex_hash = weird_hex_encoding(&hash);
-
-        let url = format!("https://sessionserver.mojang.com/session/minecraft/hasJoined?username={username}&serverId={hex_hash}&ip={}", remote_addr.ip());
+        let hex_hash = auth_digest(&hash);
+
+        let url = C::format_session_server_url(
+            server.config(),
+            server,
+            &username,
+            &hex_hash,
+            &remote_addr.ip(),
+        )
+        .to_string();
```

Comment:
`to_string` is not needed because `get` can take a `&str`. (Also you would want to use `into_owned` since that avoids an unnecessary clone)

Reply:
`to_string` and `to_owned` do the same thing, but `to_string` is more clear.
https://users.rust-lang.org/t/to-string-vs-to-owned-for-string-literals/1441/5

Reply:
I was actually referring to [`into_owned`](https://doc.rust-lang.org/stable/std/borrow/enum.Cow.html#method.into_owned), not `to_owned`.